### DISCUSSION TITLE
Implement __eq__ for most classes

### DISCUSSION THREAD

**Discussion by th1j5** [networkx/networkx#4746] [Ideas]
I wondered if it would make more sense to implement `__eq__` for some classes (encountered this mainly when I wanted to do:
```python
>>> g = nx.Graph()
>>> h = nx.Graph()
>>> g == h
False
```
This breaks some expectations when thinking about [Equality vs Identity](https://www.ecosia.org/search?q=python+equality+vs+identity&addon=opensearch), and I think networkx now just uses the default behavior, which is falling back from `==` to `is`.

My suggestion is to **either** throw an error (e.g. `NotImplementedError`) when using `==` to make sure the user understands that actually just the `id()`s are compared (and should explicitly show this by using `is`) **or** to implement them according to user expectance.

For `nx.Graph()`, this issue is relevant: #2765, but I would not use isomorphism for this (as discussed there), but more something like here: https://github.com/networkx/networkx/blob/643f45705eb68068e430367847829a35152f0ca5/networkx/testing/utils.py#L62-L65

**Comment by dschult** [comment]
Historically we punted on a definition of `__eq__` because there are too many different expectations for what `==` should do.
Unless you are basing your expectations on an application using graphs, and not based on some sense of what it "should be", I encourage you to be careful about what you expect. Just as with floating point numbers, different people expect different outcomes. 

Perhaps raising an exception explaining that you should use `is` instead of `==` is the way to go. But I expect that has rarely been a problem --  I'm pretty sure that people don't expect `==` to work with graphs -- what should that mean? And if they do, then they figure out quickly that they don't really mean what they thought they meant. This is a good lesson to learn.

For example, the `almost_graph_equals` code seems perfectly safe -- but if you have floating point values for your node attributes, or if your nodes aren't comparable then it won't be what you want.

**Reply by th1j5** [reply]
What mostly triggered this problem for me, was this:
```python
>>> g = nx.DiGraph() # or any graph, filled or not
>>> g == g.copy()
False
```
and also that I wanted to add a graph to the 'basegraph' and later subtract it again and then compare if it's still the same. (was to test some networkx behaviour) This was using `compose()` and in the end `remove_from_nodes()`.
But indeed, it might break expectancy when using floating numbers etc.
Although the copy example *did* surprise me!
on this note, I thought that nodes should be hashable, but after looking at the docs again, this is indeed not the case for node attributes, which might give problems in comparison

**Reply by th1j5** [reply]
Just my 2c as a networkx user, I don't think I'll have more to add than what I already said

**Comment by rossbar** [comment]
> I encourage you to be careful about what you expect.

In principle this "encouragement" could be made more explicit. Of the options discussed above, the one that makes the most sense to me is raising `NotImplementedError` (or similar) with a detailed message including why it's not implemented and some alternatives to achieve various likely-intended behaviors. However, explicitly defining `__eq__` would also require defining a `__hash__` method for the Graph classes, otherwise the Graph instances themselves [would not be hashable](https://docs.python.org/3/reference/datamodel.html#object.__hash__). I'm not sure the benefits would be worth the extra complication, especially since the current behavior is consistent with the default for user-defined classes.

**Reply by dschult** [reply]
Raising an exception in `__eq__` makes sense to me and it's a way to make the treatment more explicit/less confusing.
And I see what you mean about this messing up a graph being hashable. Does creating a `__hash__` function actually solve the problem?  Doesn't a dict lookup follow a hash match with a call to `__eq__`?  

Could the `__hash__` function be simply `return id(self)`?  I don't know the full implications of these kind of changes.

**Reply by rossbar** [reply]
> I don't know the full implications of these kind of changes.

I'm not too familiar with this portion of the object model either. I think `id(self)` is fine in that `id` returns the object's memory address, which should always be unique (I think?) thus is suitable for `__hash__`.

**Reply by dschult** [reply]
OK... good... So, the current (default) object `__eq__` is just `id(self)` and we already have what we want... 
It doesn't return a friendly exception. But it does what python classes do as the default.
I'm +1 for that. :}
